mypy/checkexpr.py

```diff
@@ -635,44 +633,50 @@ def check_typeddict_call(
                 # ex: Point(dict(x=42, y=1337))
                 return self.check_typeddict_call_with_dict(callee, unique_arg.analyzed, context)
 
-        if len(args) == 0:
-            # ex: EmptyDict()
-            return self.check_typeddict_call_with_kwargs(callee, {}, context)
-
         self.chk.fail(message_registry.INVALID_TYPEDDICT_ARGS, context)
         return AnyType(TypeOfAny.from_error)
 
-    def validate_typeddict_kwargs(self, kwargs: DictExpr) -> "Optional[Dict[str, Expression]]":
-        item_args = [item[1] for item in kwargs.items]
-
-        item_names = []  # List[str]
+    def validate_typeddict_kwargs(
+        self, kwargs: DictExpr
+    ) -> Optional[List[Tuple[Optional[str], Expression]]]:
+        """Validate kwargs for TypedDict constructor, e.g. Point({'x': 1, 'y': 2}).
+        Check that all items have string literal keys or are using unpack operator (**)
+        """
+        items: List[Tuple[Optional[str], Expression]] = []
         for item_name_expr, item_arg in kwargs.items:
+            # If unpack operator (**) was used, name will be None
+            if item_name_expr is None:
+                items.append((None, item_arg))
+                continue
             literal_value = None
-            if item_name_expr:
-                key_type = self.accept(item_name_expr)
-                values = try_getting_str_literals(item_name_expr, key_type)
-                if values and len(values) == 1:
-                    literal_value = values[0]
+            key_type = self.accept(item_name_expr)
+            values = try_getting_str_literals(item_name_expr, key_type)
+            if values and len(values) == 1:
+                literal_value = values[0]
             if literal_value is None:
                 key_context = item_name_expr or item_arg
                 self.chk.fail(message_registry.TYPEDDICT_KEY_MUST_BE_STRING_LITERAL, key_context)
                 return None
             else:
-                item_names.append(literal_value)
-        return dict(zip(item_names, item_args))
+                items.append((literal_value, item_arg))
+        return items
 
-    def match_typeddict_call_with_dict(
-        self, callee: TypedDictType, kwargs: DictExpr, context: Context
-    ) -> bool:
+    def match_typeddict_call_with_dict(self, callee: TypedDictType, kwargs: DictExpr) -> bool:
+        """Check that kwargs is valid set of TypedDict items, contains all required keys of callee, and has no extraneous keys"""
         validated_kwargs = self.validate_typeddict_kwargs(kwargs=kwargs)
         if validated_kwargs is not None:
-            return callee.required_keys <= set(validated_kwargs.keys()) <= set(callee.items.keys())
+            return (
+                callee.required_keys
+                <= set(dict(validated_kwargs).keys())
```

Comment:
I think that `keys()` might be enough here 🤔 

```python
>>> set() <= {'a': 1}.keys()
True
```

Reply:
You're right! Interesting to see that `dict_keys` behaves like a set. Changed!